### PR TITLE
issue #8892 - added http status code "428 precondition required".

### LINKED ISSUE #8892: Support additional HTTP response statuses
Playframework should support additional HTTP status codes, such as 428 Precondition Required, as specified in https://tools.ietf.org/html/rfc6585#section-3

(I know I can manually return any status code)

### Play Version (2.5.x / etc)
2.6

### API (Scala / Java / Neither / Both)
Both

### PR DESCRIPTION
**Pull request by gabbah**
Fixes #8892 

## Background Context
Was missing http status code 428 precondition required (Can be used in conjunction with the already existing 412 precondition failed).

### CODE REVIEW

File: framework/src/play/src/main/java/play/mvc/Http.java [hunk 1]
```diff
@@ -3325,6 +3325,7 @@ default Cookie get(String name) {
         int LOCKED = 423;
         int FAILED_DEPENDENCY = 424;
         int UPGRADE_REQUIRED = 426;
+        int PRECONDITION_REQUIRED = 428;
```
Comment (wsargent) [inline framework/src/play/src/main/java/play/mvc/Http.java]: Needs a corresponding method in Results.java, i.e. 

https://github.com/playframework/playframework/blob/843b8c5be42e671e82f0401e2efbc5c2e5f7e5f0/framework/src/play/src/main/java/play/mvc/Results.java#L1625

File: framework/src/play/src/main/scala/play/api/http/StandardValues.scala [hunk 2]
```diff
@@ -217,6 +217,7 @@ trait Status {
   val LOCKED = 423
   val FAILED_DEPENDENCY = 424
   val UPGRADE_REQUIRED = 426
+  val PRECONDITION_REQUIRED = 428
```
Comment (wsargent) [inline framework/src/play/src/main/scala/play/api/http/StandardValues.scala]: I think this needs to include the associated case class, i.e. PreconditionRequired() in Results.scala

https://github.com/playframework/playframework/blob/843b8c5be42e671e82f0401e2efbc5c2e5f7e5f0/framework/src/play/src/main/scala/play/api/mvc/Results.scala#L652

**Comment by marcospereira** [comment]
Hi @gabbah, I think that after following @wsargent suggestions we can then merge this. If you don't have time for that, let me know so that I can help you with it.

Best.

**Comment by gabbah** [comment]
> Hi @gabbah, I think that after following @wsargent suggestions we can then merge this. If you don't have time for that, let me know so that I can help you with it.
> 
> Best.

That would be great! 👍

**Comment by marcospereira** [comment]
@wsargent this is ready to be reviewed again. :-)

**Comment by renatocaval** [comment]
> Minor typo and a bit sad for the 600+ LoC change. Otherwise LGTM

The 600+ are pure repetitive code offering different variants on how to build a `Result`. This could be replaced by a Result builder in which we fix the status code first and then choose how to build the rest by selecting methods in a fluent API. 

Anyway, I don't think is something worth investing right now, but could be a future improvement.

File: framework/src/play/src/main/scala/play/api/mvc/Results.scala [hunk 3]
```diff
@@ -648,9 +648,15 @@ trait Results {
   /** Generates a ‘424 FAILED_DEPENDENCY’ result. */
   val FailedDependency = new Status(FAILED_DEPENDENCY)
 
+  /** Generates a ‘428 PRECONDITION_REQUIRED’ result. */
+  val PreconditionRequire = new Status(PRECONDITION_REQUIRED)
```
[the remark this answers is not in the document]
Comment (ignasi35) [reply framework/src/play/src/main/scala/play/api/mvc/Results.scala]: @marcospereira this suggestion is the only missing bit, I think.

**Comment by wsargent** [comment]
@renatocaval @ignasi35 the source code for `Result.java` is generated from https://github.com/jroper/play-source-generator  https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/java/play/mvc/Results.java#L257

**Comment by ignasi35** [comment]
> @renatocaval @ignasi35 the source code for Result.java is generated from ...

That makes more sense. Thanks for clarifying.

**Comment by wsargent** [comment]
approved modulo "PreconditionRequire"

**Comment by renatocaval** [comment]
Backported to 2.7.x (71556639b)

**Pull request merged**